### ISSUE TITLE
autoxtrabackup --prepare  - Select option 3 without preparing backup

### ISSUE THREAD

**Issue by Komport** [ShahriyarR/MySQL-AutoXtraBackup#158]
Hi. 
I have little confusion with option 3 in prepare. May be it is the way it should be. But I hope you will clarify.

To reproduce.

Existing databases
```
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
```
Take first backup
```
autoxtrabackup --backup
```
Create new DB
```
mysql> create database test1;
Query OK, 1 row affected (0.00 sec)
```
Take backup again
```
[root@server1 2017-11-03_00-31-49]# autoxtrabackup --backup
mysqladmin: [Warning] Using a password on the command line interface can be insecure.
```
Drop database
```
mysql> drop database test1;
Query OK, 0 rows affected (0.37 sec)

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
5 rows in set (0.23 sec)
```
And run autoxtrabackup as below.
```
mysql> drop database test1;
Query OK, 0 rows affected (0.37 sec)

mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
5 rows in set (0.23 sec)
```
This procedure will finish with success. Mysql will stop and start. But when i check databases test1 database not there. What will 3 option do in case of no prepared backups?
```
mysql> show databases;
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
Connection id:    3
Current database: *** NONE ***

+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
5 rows in set (0.11 sec)
```

**Comment by ShahriyarR** [comment]
Hi, could you please repeat the process with following command?
`autoxtrabackup -v -l DEBUG` 

I am curious about output of running commands.

**Comment by ShahriyarR** [comment]
Oh, now got your question.
The thing is, maybe I have prepared backup 1 hour ago and now want to copy-back.
So it is going to copy-back already prepared backup.

**Comment by Komport** [comment]
You can find output as below. 
```
[root@server1 2017-11-03_00-31-49]# autoxtrabackup --prepare -v -l DEBUG
2017-11-03 00:59:38 DEBUG    <pid.PidFile object at 0x7fde7cface08> entering setup
2017-11-03 00:59:38 DEBUG    <pid.PidFile object at 0x7fde7cface08> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 00:59:38 DEBUG    <pid.PidFile object at 0x7fde7cface08> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
2017-11-03 00:59:42 DEBUG    ####################################################################################################
2017-11-03 00:59:42 DEBUG    Shutting Down MySQL server: - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 00:59:42 DEBUG    ####################################################################################################
2017-11-03 00:59:49 DEBUG    
2017-11-03 00:59:49 DEBUG    ####################################################################################################
2017-11-03 00:59:49 DEBUG    Moving MySQL datadir to /home/shahriyar.rzaev/XB_TEST/mysql_datadirs  - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 00:59:49 DEBUG    ####################################################################################################
2017-11-03 00:59:52 DEBUG    Emptied /home/shahriyar.rzaev/XB_TEST/mysql_datadirs directory ...
2017-11-03 00:59:56 DEBUG    Moved datadir to /home/shahriyar.rzaev/XB_TEST/mysql_datadirs ...
2017-11-03 00:59:56 DEBUG    Creating an empty data directory ...
2017-11-03 00:59:56 DEBUG    Datadir is Created! ...
2017-11-03 00:59:56 DEBUG    ####################################################################################################
2017-11-03 00:59:56 DEBUG    Copying Back Already Prepared Final Backup: - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 00:59:56 DEBUG    ####################################################################################################
2017-11-03 01:00:04 DEBUG    ####################################################################################################
2017-11-03 01:00:04 DEBUG    Data copied back successfully! - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - #
2017-11-03 01:00:04 DEBUG    ####################################################################################################
2017-11-03 01:00:07 DEBUG    ####################################################################################################
2017-11-03 01:00:07 DEBUG    New copied-back data now owned by specified user! - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 01:00:07 DEBUG    ####################################################################################################
2017-11-03 01:00:07 DEBUG    ####################################################################################################
2017-11-03 01:00:07 DEBUG    Starting MySQL/MariaDB server: 
2017-11-03 01:00:07 DEBUG    ####################################################################################################
2017-11-03 01:00:19 DEBUG    Starting MySQL ...
2017-11-03 01:00:19 DEBUG    
2017-11-03 01:00:19 DEBUG    ####################################################################################################
2017-11-03 01:00:19 DEBUG    All data copied back successfully. 
2017-11-03 01:00:19 DEBUG    Your MySQL server is UP again
2017-11-03 01:00:19 DEBUG    ####################################################################################################
2017-11-03 01:00:19 DEBUG    <pid.PidFile object at 0x7fde7cface08> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 01:00:19 DEBUG    <pid.PidFile object at 0x7fde7cface08> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid

```

**Comment by Komport** [comment]
Another test case with empty backupdir. 
```
[root@server1 backup_dir]# autoxtrabackup --prepare -v -l DEBUG
2017-11-03 01:03:08 DEBUG    <pid.PidFile object at 0x7f2d1b6cde08> entering setup
2017-11-03 01:03:08 DEBUG    <pid.PidFile object at 0x7f2d1b6cde08> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 01:03:08 DEBUG    <pid.PidFile object at 0x7f2d1b6cde08> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
2017-11-03 01:03:13 DEBUG    ####################################################################################################
2017-11-03 01:03:13 DEBUG    Shutting Down MySQL server: - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 01:03:13 DEBUG    ####################################################################################################
2017-11-03 01:03:20 DEBUG    
2017-11-03 01:03:20 DEBUG    ####################################################################################################
2017-11-03 01:03:20 DEBUG    Moving MySQL datadir to /home/shahriyar.rzaev/XB_TEST/mysql_datadirs  - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 01:03:20 DEBUG    ####################################################################################################
2017-11-03 01:03:23 DEBUG    Emptied /home/shahriyar.rzaev/XB_TEST/mysql_datadirs directory ...
2017-11-03 01:03:25 DEBUG    Moved datadir to /home/shahriyar.rzaev/XB_TEST/mysql_datadirs ...
2017-11-03 01:03:25 DEBUG    Creating an empty data directory ...
2017-11-03 01:03:25 DEBUG    Datadir is Created! ...
2017-11-03 01:03:25 DEBUG    ####################################################################################################
2017-11-03 01:03:25 DEBUG    Copying Back Already Prepared Final Backup: - - - - - - - - - - - - - - - - - - - - - - - - - - - -#
2017-11-03 01:03:25 DEBUG    ####################################################################################################
2017-11-03 01:03:29 ERROR    Error occurred while copying back data!
2017-11-03 01:03:29 ERROR    /usr/bin/xtrabackup version 2.4.8 based on MySQL server 5.7.13 Linux (x86_64) (revision id: 97330f7)
xtrabackup: Can't change dir to '/home/shahriyar.rzaev/XB_TEST/backup_dir/full/0/' (Errcode: 2 - No such file or directory)
cannot my_setwd /home/shahriyar.rzaev/XB_TEST/backup_dir/full/0/
2017-11-03 01:03:29 ERROR    Error Occurred!
2017-11-03 01:03:29 DEBUG    <pid.PidFile object at 0x7f2d1b6cde08> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 01:03:29 DEBUG    <pid.PidFile object at 0x7f2d1b6cde08> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
[root@server1 backup_dir]# 
```

**Comment by Komport** [comment]
May be from here?

```
def recent_full_backup_file(self):
        # Return last full backup dir name

        if len(os.listdir(self.full_dir)) > 0:
            return max(os.listdir(self.full_dir))
        return 0
```

**Comment by ShahriyarR** [comment]
Yes, from our discussions I have concluded followings:
* Prevent running/choosing option 3 if backups are not prepared
* Prevent running if there is no backup

So confirmed as true bug.
Thank you very much.

**Comment by ShahriyarR** [comment]
I think here first of all, the best not to return 0 but raise a RuntimeError such as:

```
def recent_full_backup_file(self):
        # Return last full backup dir name

        if len(os.listdir(self.full_dir)) > 0:
            return max(os.listdir(self.full_dir))
        else:
            raise RuntimeError("The empty full backup directory")
```

After that handle in other places.

**Comment by ShahriyarR** [comment]
According to check if backup is prepared or not.
There is a file inside backup directory called `xtrabackup_checkpoints`. Which as backup_type column.
So in general we can check if it is equal to full-prepared, if yes then proceed. If not then raise an error.

```
$ cat xtrabackup_checkpoints
backup_type = full-prepared
from_lsn = 0
to_lsn = 2547842
last_lsn = 2547851
compact = 0
recover_binlog_info = 0
```

**Comment by ShahriyarR** [comment]
Results of test, if full backup directory is empty:

```
2017-11-03 11:31:15 DEBUG    <pid.PidFile object at 0x7f49b044b0e8> entering setup
2017-11-03 11:31:15 DEBUG    <pid.PidFile object at 0x7f49b044b0e8> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 11:31:15 DEBUG    <pid.PidFile object at 0x7f49b044b0e8> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3
3

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2017-11-03 11:31:20 DEBUG    <pid.PidFile object at 0x7f49b044b0e8> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
Traceback (most recent call last):
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/bin/autoxtrabackup", line 11, in <module>
    load_entry_point('mysql-autoxtrabackup==1.4.9', 'console_scripts', 'autoxtrabackup')()
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/autoxtrabackup.py", line 188, in all_procedure
    a.prepare_backup_and_copy_back()
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 735, in prepare_backup_and_copy_back
    self.copy_back_action()
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 696, in copy_back_action
    if self.check_if_backup_prepared(self.full_dir, self.recent_full_backup_file()):
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 34, in recent_full_backup_file
    raise RuntimeError("The full backup directory is empty, it seems you have not backups")
RuntimeError: The full backup directory is empty, it seems you have not backups
2017-11-03 11:31:20 DEBUG    <pid.PidFile object at 0x7f49b044b0e8> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
```

**Comment by ShahriyarR** [comment]
If backup is not prepared:

```
2017-11-03 11:37:50 DEBUG    <pid.PidFile object at 0x7f7de0c7a0e8> entering setup
2017-11-03 11:37:50 DEBUG    <pid.PidFile object at 0x7f7de0c7a0e8> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 11:37:50 DEBUG    <pid.PidFile object at 0x7f7de0c7a0e8> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3
3

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2017-11-03 11:37:55 ERROR    This full backup is not fully prepared, not doing copy-back!
2017-11-03 11:37:55 ERROR    Aborting!
2017-11-03 11:37:55 DEBUG    <pid.PidFile object at 0x7f7de0c7a0e8> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
Traceback (most recent call last):
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/bin/autoxtrabackup", line 11, in <module>
    load_entry_point('mysql-autoxtrabackup==1.4.9', 'console_scripts', 'autoxtrabackup')()
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/shahriyar.rzaev/virtualenvs/py_3_5_3_autoxtrabackup/lib/python3.5/site-packages/click-6.7-py3.5.egg/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/autoxtrabackup.py", line 188, in all_procedure
    a.prepare_backup_and_copy_back()
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 735, in prepare_backup_and_copy_back
    self.copy_back_action()
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 696, in copy_back_action
    if self.check_if_backup_prepared(self.full_dir, self.recent_full_backup_file()):
  File "/home/shahriyar.rzaev/MySQL-AutoXtraBackup/backup_prepare/prepare.py", line 666, in check_if_backup_prepared
    raise RuntimeError("This full backup is not fully prepared, not doing copy-back!")
RuntimeError: This full backup is not fully prepared, not doing copy-back!
2017-11-03 11:37:55 DEBUG    <pid.PidFile object at 0x7f7de0c7a0e8> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
```

**Comment by ShahriyarR** [comment]
@Komport 
Could you please test also?
The branch is `bugfix_issue158`

**Comment by ShahriyarR** [comment]
PR is created waiting for original report for any other problems on fix.

**Comment by Komport** [comment]
It is ok now. But, can we only throw an error message and not full trackback?

**Comment by ShahriyarR** [comment]
Sure, is it ok now?

```
2017-11-03 18:16:00 DEBUG    <pid.PidFile object at 0x7fa414368188> entering setup
2017-11-03 18:16:00 DEBUG    <pid.PidFile object at 0x7fa414368188> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 18:16:00 DEBUG    <pid.PidFile object at 0x7fa414368188> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3
3

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2017-11-03 18:16:06 ERROR    RuntimeError: This full backup is not fully prepared, not doing copy-back!
2017-11-03 18:16:06 DEBUG    <pid.PidFile object at 0x7fa414368188> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 18:16:06 DEBUG    <pid.PidFile object at 0x7fa414368188> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
```

```
2017-11-03 18:17:17 DEBUG    <pid.PidFile object at 0x7f6157493188> entering setup
2017-11-03 18:17:17 DEBUG    <pid.PidFile object at 0x7f6157493188> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 18:17:17 DEBUG    <pid.PidFile object at 0x7f6157493188> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 3
3

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2017-11-03 18:17:22 ERROR    RuntimeError: The full backup directory is empty, it seems you have not backups
2017-11-03 18:17:22 DEBUG    <pid.PidFile object at 0x7f6157493188> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-03 18:17:22 DEBUG    <pid.PidFile object at 0x7f6157493188> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
```
```

**Comment by Komport** [comment]
Great 👍

**Comment by ShahriyarR** [comment]
Good, then it can be merged to `release_v1.5.0` and it will be available on 1.5.0 version of autoxtrabackup

**Comment by ShahriyarR** [comment]
Merged.
